include/deal.II/lac/cuda_solver_direct.h

```diff
@@ -0,0 +1,114 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_cuda_solver_direct_h
+#define dealii_cuda_solver_direct_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_CUDA
+#include <deal.II/base/cuda.h>
+#include <deal.II/lac/cuda_sparse_matrix.h>
+#include <deal.II/lac/cuda_vector.h>
+#include <deal.II/lac/solver_control.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace CUDAWrappers
+{
+  /**
+   * Direct solvers. These solvers call cuSOLVER underneath.
+   *
+   * @note Instantiations for this template are provided for <tt>@<float@> and
+   * @<double@></tt>.
+   *
+   * @ingroup CUDAWrappers
+   * @author Bruno Turcksin
+   * @date 2018
+   */
+  template <typename Number>
+  class SolverDirect
+  {
+  public:
+    struct AdditionalData
+    {
+      /**
+       * Set the additional data field to the desired solver.
+       */
+      explicit
+      AdditionalData (const std::string &solver_type = "LU_dense");
+
+      /**
+       * Set the solver type. Possibilities are:
+       * <ul>
+       * <li> "Cholesky" which uses Cholesky on the device </li>
+       * <li> "LU_dense" which converts the sparse matrix to a dense matrix and
+       * uses LU factorization </li>
+       * <li> "LU_host" which uses LU factorization on the host </li>
```

Comment:
Out of curiosity, and before reading the full PR to the end: What does "LU factorization on the host" mean technically? Does CUDA include a sparse LU factorization package? Is it one of the well-known ones like Pardiso or UMFPACK?

Reply:
This means that the data is moved to the host where the factorization is done. Unfortunately, there is no function to do a LU decomposition on the device if the matrix is sparse. Like the two other functions this function uses `cuSOLVER`.